#' Rscript --vanilla get_frame0_ratio.R input.tsv txinfo.tsv appris_tsl.tsv output.tsv
#' @path_input: input bam.tsv path (generated by `samtools view xxx.Aligned.toTranscriptome.out.bam | cut -f1,3,4,6`)
#' @path_txinfo: input txinfo path (human:/nfs_data/database/ref_genomes/human_GRCh38p13/ens104/txinfo_human_ens104.tsv;mouse:/nfs_data/database/ref_genomes/mouse_GRCm39/ens106/txinfo_mouse_ens106.tsv)
#' @path_appris_tsl: input appris_tsl path
#' @path_output: output path


library(tidyverse)

args <- commandArgs(trailingOnly = TRUE)
path_input <- args[1]
path_txinfo <- args[2]
path_appris_tsl <- args[3]
path_output <- args[4]

appris_tsl <- read_tsv(path_appris_tsl)
names(appris_tsl) <- c('gene_id', 'tx_name', 'appris', 'tsl')
appris_tsl <- appris_tsl |>
  mutate(appris = factor(appris, levels = c(
    c("principal1", "principal2", "principal3", "principal4", "principal5",
      "alternative1", "alternative2", ""))))

txinfo <- read_tsv(path_txinfo)
tx_bam <- read_tsv(path_input, col_names = F) |>
  select(seq_id = X1, tx_name = X2, pos = X3, matched = X4) |>
  mutate(match_len = str_extract(matched, '[0-9]*(?=M)'), 
         match_len = as.integer(match_len)) |>
  left_join(txinfo[, c(1, 6, 7)], by = 'tx_name') |>
  left_join(appris_tsl[, 2:3], by = 'tx_name')

## total
len_total <- tx_bam |>
  distinct(seq_id, match_len) |>
  count(match_len, name = 'total')

## use matched len
tx_filter_matched <- tx_bam |> 
  filter(pos >= utr5_len + 1, 
         pos + match_len - 1 <= utr5_len + cds_len) |>
  mutate(dis = (pos - utr5_len -  1)%%3) |>
  group_by(seq_id) |>
  arrange(appris) |>
  filter(!duplicated(seq_id)) |>
  ungroup()

match_len_ratio <- tx_filter_matched |>
  count(match_len, dis) |>
  mutate(dis = case_when(
    dis == 0 ~ 'f0', 
    dis == 1 ~ 'f1', 
    dis == 2 ~ 'f2'
  )) |>
  pivot_wider(names_from = dis, values_from = n) |>
  rowwise() |>
  mutate(f0_percent = f0/sum(f0, f1, f2), 
         f1_percent = f1/sum(f0, f1, f2), 
         f2_percent = f2/sum(f0, f1, f2), 
         max_ratio = max(f0_percent, f1_percent, f2_percent), 
         cds_total = sum(f0, f1, f2))

## merge
merge <- match_len_ratio |>
  full_join(len_total, by = 'match_len')

write_tsv(merge, file = path_output)
